def.cf

```diff
@@ -4,30 +4,95 @@
 #  - common/global variables and classes here
 #
 ###############################################################################
+body file control
+{
+      inputs => { @(def.extra_inputs) };
+}
 
 bundle common def
 
 {
+  classes:
+      # all extra classes are defined true
+      "$(extra_classes)" expression => "any", meta => { "extra" };
+
+      # all override classes are defined true
+      "$(override_classes)" expression => "any", meta => { "override" };
+
+      "have_augments_file" expression => fileexists($(augments_file)), scope => "bundle";
+
+      "augments_are_json" expression => regcmp(".+\.json", $(augments_file)), scope => "bundle";
+      "have_json_augments" and => { "have_augments_file", "augments_are_json" }, scope => "bundle";
+
+      "augments_are_yaml" expression => regcmp(".+\.yaml", $(augments_file)), scope => "bundle";
+      "have_yaml_augments" and => { "have_augments_file", "augments_are_yaml" }, scope => "bundle";
+
+      "have_roles" expression => isvariable("augments[roles]"), scope => "bundle";
+      "have_inputs" expression => isvariable("augments[inputs]"), scope => "bundle";
+
+    have_roles::
+      "$(extra_roles[byname][$(roles_byname_keys)])" expression => classmatch("$(roles_byname_keys)"), meta => { "role" };
+      "$(roles_byrole_keys)" expression => classmatch("$(extra_roles[byrole][$(roles_byrole_keys)])"), meta => { "role" };
 
   vars:
 
     any::
+      "augments_file" string => ifelse(fileexists("$(this.promise_dirname)/services/def.yaml"), "$(this.promise_dirname)/services/def.yaml",
+                                      "$(this.promise_dirname)/services/def.json");
+
+      "defvars" slist => variablesmatching("default:def\..*", "defvar");
+
+      "extra_classes" slist => readstringlist("$(this.promise_dirname)/services/extra_classes.txt", "#.*", "\s", 10k, 100k);
+
+    have_augments_file::
+      "augments" data => readjson($(augments_file), 100k), ifvarclass => "augments_are_json";
+      "augments" data => readyaml($(augments_file), 100k), ifvarclass => "augments_are_yaml";
```

Comment:
We cant have yamlstuff until 3.8 since it breaks compatibility with 3.6. I think it's fine to have it there but commented out.